Core/MIPS/MIPSVFPUUtils.cpp

```diff
@@ -711,6 +712,177 @@ float vfpu_dot(const float a[4], const float b[4]) {
+	// NOTE: the cast for 1<<31 produces the correct value,
+	// since the _mm_cvttps_epi32 error code just happens
+	// to be 0x80000000.
+	// So (since we pre-multiplied m by 2), we need
+	// (m>>1)>>(mexp-exps),
+	// i.e. m>>(mexp+1-exps),
+	// i.e. (m<<(32-(mexp+1-exps)))>>32,
+	// i.e. (m<<(exps-(mexp-31)))>>32.
+	__m128i amounts = _mm_sub_epi32(exps, _mm_set1_epi32(mexp - 31));
+	// Clamp by 0. Both zero and negative amounts produce zero,
+	// since they correspond to right-shifting by 32 or more bits.
+	amounts = _mm_and_si128(amounts, _mm_cmpgt_epi32(amounts, _mm_set1_epi32(0)));
+	// Set up multipliers.
+	__m128i bits = _mm_add_epi32(_mm_set1_epi32(0x3F800000), _mm_slli_epi32(amounts, 23));
+	__m128i muls = _mm_cvttps_epi32(_mm_castsi128_ps(bits));
+	m = mulhi32x4(m, muls);
+	// Extract signs.
+	__m128i signs = _mm_cmpgt_epi32(
+			_mm_set1_epi32(0),
+			_mm_xor_si128(_mm_castps_si128(A), _mm_castps_si128(B)));
+	// Apply signs to m.
+	m = _mm_sub_epi32(_mm_xor_si128(m, signs), signs);
+	// Horizontal sum.
+	// See https://stackoverflow.com/questions/6996764/fastest-way-to-do-horizontal-sse-vector-sum-or-other-reduction
+	__m128i h64 = _mm_shuffle_epi32(m, _MM_SHUFFLE(1, 0, 3, 2));
+	__m128i s64 = _mm_add_epi32(h64, m);
+	__m128i h32 = _mm_shufflelo_epi16(s64, _MM_SHUFFLE(1, 0, 3, 2));
+	__m128i s32 = _mm_add_epi32(s64, h32);
+	int32_t mant_sum = _mm_cvtsi128_si32(s32);
+
+	// The rest is scalar.
+	uint32_t sign_sum = 0;
+	if (mant_sum < 0) {
+		sign_sum = 0x80000000;
+		mant_sum = -mant_sum;
+	}
+
+	// Truncate off the extra bits now.  We want to zero them for rounding purposes.
+	mant_sum >>= EXTRA_BITS;
+
+	if (mant_sum == 0 || max_exp <= 0) {
+		return 0.0f;
+	}
+
+	if(is_default_rounding_mode)
+	{
+		float2int r;
+		r.f = (float)mant_sum;
+		mant_sum = (r.i & 0x007FFFFF) | 0x00800000;
+		max_exp += (r.i >> 23) - 0x96;
+	}
+	else
+	{
+		int8_t shift = (int8_t)clz32_nonzero(mant_sum) - 8;
+		if (shift < 0) {
+			// Round to even if we'd shift away a 0.5.
+			const uint32_t round_bit = 1 << (-shift - 1);
+			if ((mant_sum & round_bit) && (mant_sum & (round_bit << 1))) {
+				mant_sum += round_bit;
+				shift = (int8_t)clz32_nonzero(mant_sum) - 8;
+			} else if ((mant_sum & round_bit) && (mant_sum & (round_bit - 1))) {
+				mant_sum += round_bit;
+				shift = (int8_t)clz32_nonzero(mant_sum) - 8;
+			}
+			mant_sum >>= -shift;
+			max_exp += -shift;
+		} else {
+			mant_sum <<= shift;
+			max_exp -= shift;
+		}
+		_dbg_assert_msg_((mant_sum & 0x00800000) != 0, "Mantissa wrong: %08x", mant_sum);
+	}
+
+	if (max_exp >= 255) {
+		max_exp = 255;
+		mant_sum = 0;
+	} else if (max_exp <= 0) {
+		return 0.0f;
+	}
+
+	float2int result;
+	result.i = sign_sum | (max_exp << 23) | (mant_sum & 0x007FFFFF);
+	return result.f;
+}
+
+#endif // defined(__SSE2__)
+
+float vfpu_dot(const float a[4], const float b[4]) {
+#if defined(__SSE2__)
+	return vfpu_dot_sse2(a, b);
+#endif
+	return vfpu_dot_cpp(a, b);
+}
```

Comment:
please do an #else, if nothing else for clarity.

Reply:
Done.